twoliter/src/lock.rs

```diff
@@ -28,3 +38,230 @@ pub(crate) struct LockedImage {
     /// The resolved image uri of the dependency
     pub source: String,
 }
+
+impl LockedImage {
+    pub fn new(vendor: &Vendor, image: &Image) -> Self {
+        Self {
+            name: image.name.clone(),
+            version: image.version.clone(),
+            vendor: image.vendor.clone(),
+            source: format!("{}/{}:v{}", vendor.registry, image.name, image.version),
+        }
+    }
+}
+
+/// The hash should not contain the source to allow for collision detection
+impl Hash for LockedImage {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
+        self.version.hash(state);
+        self.vendor.hash(state);
+    }
+}
+
+#[derive(Deserialize, Debug)]
+#[allow(dead_code)]
+struct ImageMetadata {
+    /// The name of the kit
+    pub name: String,
+    /// The version of the kit
+    pub version: Version,
+    /// The required sdk of the kit,
+    pub sdk: Image,
+    /// Any dependent kits
+    pub kit: Vec<Image>,
+}
+
+#[derive(Deserialize, Debug)]
+#[allow(dead_code)]
+struct ManifestListView {
+    manifests: Vec<ManifestView>,
+}
+
+#[derive(Deserialize, Debug)]
+#[allow(dead_code)]
+struct ManifestView {
+    digest: String,
+}
+
+macro_rules! docker {
+    ($arg: expr, $error_msg: expr) => {{
+        let output = Command::new("docker")
+            .args($arg)
+            .output()
+            .await
+            .context($error_msg)?;
+        ensure!(output.status.success(), $error_msg);
+        output.stdout
+    }};
+}
+
+#[allow(dead_code)]
+impl Lock {
+    pub(crate) async fn load(project: &Project) -> Result<Self> {
+        let lock_file_path = project.project_dir().join("Twoliter.lock");
+        if lock_file_path.exists() {
+            let lock_str = read_to_string(&lock_file_path)
+                .await
+                .context("failed to read lockfile")?;
+            let lock: Self =
+                toml::from_str(lock_str.as_str()).context("failed to deserialize lockfile")?;
+            // The digests must match, if changes are needed twoliter
+            ensure!(lock.digest == project.digest()?, "changes have occured to Twoliter.toml that require an update to Twoliter.lock, if intentional please run twoliter update");
+            return Ok(lock);
+        }
+        Self::create(project).await
+    }
+
+    pub(crate) async fn create(project: &Project) -> Result<Self> {
+        let lock_file_path = project.project_dir().join("Twoliter.lock");
```

Comment:
nit: can make this a constant:
```suggestion
        let lock_file_path = project.project_dir().join(TWOLITER_LOCK);
```